### ISSUE TITLE
After some time hickory-resolver lookups start timing out

### ISSUE THREAD

**Issue by conectado** [CloudFire-LLC/cloudfire#2939]
This has been seen to happen on macos and seems to happen consistently except when token expires before that happening.

**Comment by conectado** [comment]
It seems like there's an easy way to reproduce:
* Connect the client
* Disconnect and reconnect your network connection quickly
* DNS messages will start to fail

**Comment by conectado** [comment]
So the problem is that when a tunnel reconnects, after connlib is reinitialized, the only system dns resolver that's obtained is the one installed by connlib `100.111.1.1` which is useless as a fallback resolver.

**Comment by jamilbk** [comment]
Seeing similar behavior when the default route changes, such as changing service order from Wifi to Ethernet or to my mobile hotspot.

What we could do is, if DNS resolution fails or times out, call `get_system_default_resolvers` again and re-initialize the resolver if the list is different.

**Comment by jamilbk** [comment]
```
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	name_server connection failure: request timed out	12:04:39.948653-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	name_server connection failure: request timed out	12:04:39.948682-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	sending request: [Query { name: Name("optimizationguide-pa.googleapis.com"), query_type: A, query_class: IN }]	12:04:39.948808-0800
error	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	DNS lookup failed: Error with the DNS fallback lookup  err=DNSFallbackKind(Timeout)	12:04:39.948826-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	reconnecting: NameServerConfig { socket_addr: [2600:1700:3ecb:2410::1]:53, protocol: Udp, tls_dns_name: None, trust_negative_responses: true, bind_addr: None }	12:04:39.948961-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	name_server connection failure: request timed out	12:04:39.948970-0800
error	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	DNS lookup failed: Error with the DNS fallback lookup  err=DNSFallbackKind(Timeout)	12:04:39.949100-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	enqueueing message:QUERY:[Query { name: Name("optimizationguide-pa.googleapis.com"), query_type: A, query_class: IN }]	12:04:39.949152-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	reconnecting: NameServerConfig { socket_addr: 192.168.1.254:53, protocol: Udp, tls_dns_name: None, trust_negative_responses: true, bind_addr: None }	12:04:39.949275-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	enqueueing message:QUERY:[Query { name: Name("optimizationguide-pa.googleapis.com"), query_type: A, query_class: IN }]	12:04:39.949417-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	final message: ; header 20783:QUERY:RD:NoError:QUERY:0/0/0
; query
;; optimizationguide-pa.googleapis.com. IN A	12:04:39.949594-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	final message: ; header 36985:QUERY:RD:NoError:QUERY:0/0/0
; query
;; optimizationguide-pa.googleapis.com. IN A	12:04:39.949711-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	created socket successfully	12:04:39.949865-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	created socket successfully	12:04:39.950128-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	querying: 2.10.20.172.in-addr.arpa PTR	12:04:40.097645-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	sending request: [Query { name: Name("2.10.20.172.in-addr.arpa"), query_type: PTR, query_class: IN }]	12:04:40.098189-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	existing connection: NameServerConfig { socket_addr: [2600:1700:3ecb:2410::1]:53, protocol: Udp, tls_dns_name: None, trust_negative_responses: true, bind_addr: None }	12:04:40.098541-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	enqueueing message:QUERY:[Query { name: Name("2.10.20.172.in-addr.arpa"), query_type: PTR, query_class: IN }]	12:04:40.098797-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	existing connection: NameServerConfig { socket_addr: 192.168.1.254:53, protocol: Udp, tls_dns_name: None, trust_negative_responses: true, bind_addr: None }	12:04:40.099071-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	enqueueing message:QUERY:[Query { name: Name("2.10.20.172.in-addr.arpa"), query_type: PTR, query_class: IN }]	12:04:40.099303-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	final message: ; header 60812:QUERY:RD:NoError:QUERY:0/0/0
; query
;; 2.10.20.172.in-addr.arpa. IN PTR	12:04:40.099678-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	final message: ; header 34288:QUERY:RD:NoError:QUERY:0/0/0
; query
;; 2.10.20.172.in-addr.arpa. IN PTR	12:04:40.099792-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	created socket successfully	12:04:40.100167-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	created socket successfully	12:04:40.100492-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	name_server connection failure: request timed out	12:04:40.276312-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	name_server connection failure: request timed out	12:04:40.276802-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	sending request: [Query { name: Name("2.10.20.172.in-addr.arpa"), query_type: PTR, query_class: IN }]	12:04:40.277041-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	reconnecting: NameServerConfig { socket_addr: [2600:1700:3ecb:2410::1]:53, protocol: Udp, tls_dns_name: None, trust_negative_responses: true, bind_addr: None }	12:04:40.277266-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	enqueueing message:QUERY:[Query { name: Name("2.10.20.172.in-addr.arpa"), query_type: PTR, query_class: IN }]	12:04:40.277503-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	reconnecting: NameServerConfig { socket_addr: 192.168.1.254:53, protocol: Udp, tls_dns_name: None, trust_negative_responses: true, bind_addr: None }	12:04:40.277698-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	enqueueing message:QUERY:[Query { name: Name("2.10.20.172.in-addr.arpa"), query_type: PTR, query_class: IN }]	12:04:40.277923-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	final message: ; header 57513:QUERY:RD:NoError:QUERY:0/0/0
; query
;; 2.10.20.172.in-addr.arpa. IN PTR	12:04:40.278115-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	final message: ; header 59546:QUERY:RD:NoError:QUERY:0/0/0
; query
;; 2.10.20.172.in-addr.arpa. IN PTR	12:04:40.278162-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	created socket successfully	12:04:40.278528-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	created socket successfully	12:04:40.278958-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	name_server connection failure: request timed out	12:04:40.403623-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	name_server connection failure: request timed out	12:04:40.405371-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	sending request: [Query { name: Name("www.apple.com"), query_type: AAAA, query_class: IN }]	12:04:40.405709-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	reconnecting: NameServerConfig { socket_addr: [2600:1700:3ecb:2410::1]:53, protocol: Udp, tls_dns_name: None, trust_negative_responses: true, bind_addr: None }	12:04:40.406053-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	enqueueing message:QUERY:[Query { name: Name("www.apple.com"), query_type: AAAA, query_class: IN }]	12:04:40.406369-0800
debug	dev.firezone.firezone	connlib	FirezoneNetworkExtensionmacOS	reconnecting: NameServerConfig { socket_addr: 192.168.1.254:53, protocol: Udp, tls_dns_name: None, trust_negative_responses: true, bind_addr: None }	12:04:40.406672-0800
```

**Comment by conectado** [comment]
> Seeing similar behavior when the default route changes, such as changing service order from Wifi to Ethernet or to my mobile hotspot.
> 
> What we could do is, if DNS resolution fails or times out, call `get_system_default_resolvers` again and re-initialize the resolver if the list is different.

That wouldn't solve it, since the problem is that when the tunnel if reinitialized after a network change the only name server returned by `get_system_default_resolver` is our sentinel name server.

**Comment by jamilbk** [comment]
Some more interesting data:

* So macOS updates `/etc/resolv.conf` when the tunnel comes up and goes down to our sentinel `100.100.111.1`.
* `scutil --dns` does have the info we want, but sadly this API isn't available from iOS. Also note that it doesn't show the connlib DNS server as "reachable":
```
jamil@Airbook-Mac:~ % scutil --dns
DNS configuration

resolver #1
  nameserver[0] : 100.100.111.1
  if_index : 52 (utun13)
  flags    : Supplemental, Request A records, Request AAAA records
  reach    : 0x00000000 (Not Reachable)
  order    : 104400

resolver #2
  nameserver[0] : 10.128.128.128
  if_index : 12 (en0)
  flags    : Request A records
  reach    : 0x00020002 (Reachable,Directly Reachable Address)
  order    : 200000
``` 

So here's what's happening:

- `Session.connect` -> `getSystemDefaultResolvers` returns `1.1.1.1` (example)
- Change WiFi networks, or change other connectivity
- App sets tunnel into `reasserting` state, `Session.disconnect` is called, but tunnel is **not** brought down. This means `/etc/resolv.conf` still contains `100.100.111.1`.
- Internet connectivity resumes, `Session.connect` is called, but this time, `getSystemDefaultResolvers` returns `100.100.111.1`.

And here's my proposed fix:

- Before bringing the tunnel up, read the existing `/etc/resolv.conf` resolvers
- Add these to the tunnel's `dnsServers` configuration

That way, anytime connlib calls `getSystemDefaultResolvers`, it'll have whatever the system was using.

**Comment by roop** [comment]
I suspect that the API we're using relies on /etc/resolv.conf to be rewritten by the OS, and this rewriting takes time to get updated (understandably).

> scutil --dns does have the info we want, but sadly this API isn't available from iOS. Also note that it doesn't show the connlib DNS server as "reachable"

The source code for scutils is open-sourced by Apple, so we can check it's possible to integrate that:
  - https://opensource.apple.com/source/configd/configd-596.15/scutil.tproj/scutil.c.auto.html
  - https://opensource.apple.com/source/configd/configd-596.15/dnsinfo/dnsinfo_internal.h.auto.html
  - https://opensource.apple.com/source/configd/configd-1109.60.2/dnsinfo/dnsinfo_copy.c.auto.html

The ideal solution would be if we use the scutils method to find the resolver (we should be able to also exclude Firezone's resolver), and do it every time the path changes as per  NWPathMonitor.

Would also be great if connlib didn't have to be re-setup whenever the system default DNS server changes.

**Comment by jamilbk** [comment]
Fixed in #3012
